### PR TITLE
Implement Selection and Marquee Zoom

### PR DESCRIPTION
**Pull request by kaansancak**

### CODE REVIEW

**Comment by hasanbalci** [comment]
@kaansancak 
For the first feature (Zoom to selected):

- Let this feature include the edges as well, not only the nodes. When an edge is selected, we should be able to zoom to that edge. 
- Change the label of the button in the demo file as "Zoom to Selected".

For the second feature (Marquee zoom):

- There is a bug: First click the marquee zoom button. Then click or drag the mouse without pressing the shift button, graph disappears (possibly it makes zoom again)
- Currently, it zooms to the selected rectangle, but does not pan the zoomed area to the center of the viewport. (See my comment on the related part of the code)  
- If the selected rectangle does not intersect with the graph's bounding box, then ignore the marquee zoom.
- If any side of the rectangle is smaller than 100px, complete it to 100px so that the minimum area to zoom becomes at least 100*100.
-  Add the text "**Shift + drag** to specify region" next to the Marquee Zoom button in demo file.

General comments:

- Please add an option for the animation duration and use it in both features. Default can be 1500 ms.  
- Don't forget to update README file with the new features

File: src/view-utilities.js [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
+  .style()
+  .selector("node.highlighted")
+  .css(options.node.highlighted)
+  .selector("node.unhighlighted")
+  .css(options.node.unhighlighted)
+  .selector("edge.highlighted")
+  .css(options.edge.highlighted)
+  .selector("edge.unhighlighted")
+  .css(options.edge.unhighlighted)
+  .update();
```
Comment (hasanbalci) [inline src/view-utilities.js]: Please revert the changes on the indentation back here and in similar cases.

File: src/view-utilities.js [hunk 2]
```diff
@@ -114,16 +112,72 @@ var viewUtilities = function (cy, options) {
...
+        eles: eles,
+        padding: 20
+      }
+    }, {
+      duration: 1000
+    });  
+    return eles;
+  };
+
+  instance.marqueeZoom = function( canvas){
```
Comment (hasanbalci) [inline src/view-utilities.js]: Unnecessary parameter, please remove it. Also, we don't put single space before parameter names, please remove it here and in similar cases.

File: src/view-utilities.js [hunk 3]
```diff
@@ -114,16 +112,72 @@ var viewUtilities = function (cy, options) {
...
+        p_start_x = event.position.x;
+        p_start_y = event.position.y;
+      }
+    });
+
+    cy.one('tapend', function( event){
+      p_end_x = event.position.x;
+      p_end_y = event.position.y;
+      var zoomLevel = Math.min( cy.width()/ ( Math.abs(p_end_x- p_start_x)), cy.height() / Math.abs( p_end_y - p_start_y));
+      cy.animate({
```
Comment (hasanbalci) [inline src/view-utilities.js]: In order to center the selected rectangle during zoom, use pan option in the animate function instead of the position inside the zoom option. Pan option in the animate pans the top-left corner of the boundingBox of the graph to the given position value. 
Steps to be done:

-   First find the top-left point of the boundingBox of the graph. 
-   Find the distance between the center of the selected rectangle and the top-left point in both x and y coordinates.
-   Multiply these distances with zoomLevel value (Let say the resulting values are a and b (give more convenient names)).
-   In the pan option of the animation, give the position as `x: cy.width()/2 - a, y: cy.height()/2 - b`.

This should work, you do not have to give a position inside the zoom option.

File: src/view-utilities.js [hunk 4]
```diff
@@ -124,20 +125,20 @@ var viewUtilities = function (cy, options) {
...
+  instance.zoomToSelected = function (eles, zoomSpeed){
     eles.unselect();
     cy.animate({
       fit: {
         eles: eles,
         padding: 20
       }
     }, {
-      duration: 1000
+      duration: 30 * zoomSpeed //Default:1500( %50)
```
Comment (hasanbalci) [inline src/view-utilities.js]: Please add an option named as "zoomAnimationDuration" to the options object at the beginning of the index.js file, after setDisplayOnHide option, and set its value to 1500. Then call that option here as 
`duration: options.zoomAnimationDuration`
Remove zoomSpeed parameter.

File: src/view-utilities.js [hunk 5]
```diff
@@ -153,29 +154,79 @@ var viewUtilities = function (cy, options) {
...
-          level: zoomLevel}, 
-          duration: 2000,
-          complete: function() {
-            cy.autounselectify(false);
-            cy.elements().unselect();
-          }});   
-    })
+        pan : {x: (cy.width()/2 - diff_x), y: (cy.height()/2 - diff_y)},
+        zoom : {level: zoomLevel}, 
+        duration: 30 * zoomSpeed, //Default:1500( %50)
```
Comment (hasanbalci) [inline src/view-utilities.js]: Same as zoomToSelected function. Call zoomAnimationDuration option here.
Remove zoomSpeed parameter and the slider in the demo file.

File: src/view-utilities.js [hunk 6]
```diff
@@ -153,29 +154,79 @@ var viewUtilities = function (cy, options) {
...
 
-    cy.one('tapend', function( event){
-      p_end_x = event.position.x;
-      p_end_y = event.position.y;
-      var zoomLevel = Math.min( cy.width()/ ( Math.abs(p_end_x- p_start_x)), cy.height() / Math.abs( p_end_y - p_start_y));
+    cy.one('tapend', function(event){
+      rect_end_pos_x = event.position.x;
+      rect_end_pos_y = event.position.y;
+      //Find top left of rectangle
+      if( rect_start_pos_x > rect_start_pos_x){
```
Comment (hasanbalci) [inline src/view-utilities.js]: I think this line will be ` rect_start_pos_x > rect_end_pos_x`.

Moreover, here rect_start_pos_x and rect_start_pos_y values become undefined if the user click to the Marquee Zoom button and then tap to the viewport without pressing to the shift button. You should check whether these values are null or not. If they are null, then this means that user didn't press the shift button so ignore the marquee zoom.

File: src/view-utilities.js [hunk 7]
```diff
@@ -153,29 +154,79 @@ var viewUtilities = function (cy, options) {
...
+        var temp = rect_start_pos_x;
+        rect_start_pos_x = rect_end_pos_x;
+        rect_end_pos_x = temp;
+      }
+      if( rect_start_pos_y > rect_end_pos_y){
+        var temp = rect_start_pos_y;
+        rect_start_pos_y = rect_end_pos_y;
+        rect_end_pos_y = temp;
+      }
```
Comment (hasanbalci) [inline src/view-utilities.js]: I think the organization of the tapend after this point should be as folllows:

1. Check whether any side of the rectangle is less than 100. For example, if rect_start_pos_x is 40 and rect_end_pos_x is 100, then the difference is 60, so we need to expand the rectangle both from left and right side to make the distance 100. To achieve that, make the value of rect_start_pos_x 20 and the value of rect_end_pos_x 120. Similarly, if rect_start_pos_y is 210 and rect_end_pos_y is 300, make the value of rect_start_pos_y 205 and value of rect_end_pos_y 305. Do not use zoomLevel here.
At the end of this step, we will have a rectangle whose width and height values are at least 100. 
2. Check whether the resulting rectangle after step 1 has an intersection with the boundingBox of the graph. If there is no intersection, ignore marquee zoom.
3. Calculate the zoom level according to the rectangle resulting after step 1. 
4. Do the calculations to find the pan position. 
5. Call the animate function.

File: src/view-utilities.js [hunk 8]
```diff
@@ -153,29 +154,79 @@ var viewUtilities = function (cy, options) {
...
+      if((Math.min(rect_start_pos_x, rect_end_pos_x) > cy.elements().boundingBox().x2)
+        ||(Math.max(rect_start_pos_x, rect_end_pos_x) < cy.elements().boundingBox().x1)
+        ||(Math.min(rect_start_pos_y, rect_end_pos_y) > cy.elements().boundingBox().y2)
+        ||(Math.max(rect_start_pos_y, rect_end_pos_y) < cy.elements().boundingBox().y1)){
+        cy.autounselectify(false);
+        cy.elements().unselect();
+        return;        
+      }
+      //if zoom level reaches max abort marquee zoom
+      if(zoomLevel > cy.maxZoom()){
```
Comment (hasanbalci) [inline src/view-utilities.js]: Since we will have a rectangle whose width and height values are at least 100, I don't think that zoom level will reach to the maxZoom level. This may be an unnecessary check.

Comment (kaansancak) [reply src/view-utilities.js]: @hasanbalci this checks whether the selected region is a point( zoom level becomes infinity) for the bug mentioned at comments but as you said I removed it since we extend sides of the box to 100px.

File: src/view-utilities.js [hunk 9]
```diff
@@ -160,67 +160,66 @@ var viewUtilities = function (cy, options) {
...
         rect_start_pos_y = rect_end_pos_y;
         rect_end_pos_y = temp;
       }
-      //Calculate zoom level
-      var zoomLevel = Math.min( cy.width()/ ( Math.abs(rect_end_pos_x- rect_start_pos_x)), 
-        cy.height() / Math.abs( rect_end_pos_y - rect_start_pos_y));
+
+      //Extend sides of selected rectangle to 100px if less than 100px
+      if(rect_end_pos_x - rect_start_pos_x < 100){
+        var extendPx = ( 100 - (rect_end_pos_x - rect_start_pos_y)) / 2;
```
Comment (hasanbalci) [inline src/view-utilities.js]: Must be `var extendPx = ( 100 - (rect_end_pos_x - rect_start_pos_x)) / 2;`

Comment (kaansancak) [reply src/view-utilities.js]: @hasanbalci fixed.

**Comment by kaansancak** [comment]
@hasanbalci there is a bug: Select a renctangle around a single node. After first zoom, select a similar rectangle(around same node), sometimes graph disappears.

**Comment by ugurdogrusoz** [comment]
@kaansancak added this as an issue
